ci/aggregate_data.py

```diff
@@ -0,0 +1,222 @@
+    if agg["avg_count"] > 0:
+        final_values["avg"] = agg["avg_sum"] / agg["avg_count"]
+    if agg["med_count"] > 0:
+        final_values["med"] = agg["med_sum"] / agg["med_count"]
+    if agg["p90_count"] > 0:
+        final_values["p(90)"] = agg["p90_sum"] / agg["p90_count"]
+    if agg["p95_count"] > 0:
+        final_values["p(95)"] = agg["p95_sum"] / agg["p95_count"]
+    if agg["value_count"] > 0:
+        final_values["value"] = agg["value_sum"] / agg["value_count"]
+
+    # If this is a counter that might have a rate, compute it from count / total_time_s
+    # (assuming distributed tests in parallel)
+    if agg["has_rate"] and agg["has_count"] and total_time_s > 0:
+        final_values["rate"] = final_values["count"] / total_time_s
+
+    return {
+        "type": agg["type"],
+        "contains": agg["contains"],
+        "values": final_values
+    }
+
+def combine_all_docs(docs):
+    """
+    1) Find maximum .state.testRunDurationMs (assuming parallel test).
+    2) Aggregate metrics from all docs in a single pass.
+    3) Finalize the aggregates into one set of .metrics.
+    4) Return the final merged doc.
+    """
+    if not docs:
+        return {}
+
+    # 1) Determine the global "test duration" (max of testRunDurationMs).
+    max_duration_ms = 0
+    for doc in docs:
+        state = doc.get("state", {})
+        dur = state.get("testRunDurationMs", 0)
+        if dur > max_duration_ms:
+            max_duration_ms = dur
+    total_time_s = max_duration_ms / 1000.0
+
+    # 2) Build aggregators for each metric from all docs
+    metric_aggregators = {}  # metric_name -> aggregator
+    for doc in docs:
+        metrics = doc.get("metrics", {})
+        for m_name, m_data in metrics.items():
+            m_type = m_data.get("type", "")
+            m_contains = m_data.get("contains", "")
+
+            if m_name not in metric_aggregators:
+                metric_aggregators[m_name] = create_metric_aggregator(m_type, m_contains)
+
+            # Update aggregator with this doc's values
+            values_dict = m_data.get("values", {})
+            update_metric_aggregator(metric_aggregators[m_name], values_dict)
+
+    # 3) Finalize each aggregator -> produce final metrics
+    merged_metrics = {}
+    for m_name, agg in metric_aggregators.items():
+        merged_metrics[m_name] = finalize_metric_aggregator(agg, total_time_s)
+
+    # 4) Construct the final merged doc
+    # We keep root_group/options from the first doc; override testRunDurationMs with max
+    first = docs[0]
+    merged = {
+        "root_group": first.get("root_group", {}),
+        "options": first.get("options", {}),
+        "state": {
+            **first.get("state", {}),
+            "testRunDurationMs": max_duration_ms
+        },
+        "metrics": merged_metrics
+    }
+    return merged
+
+def main():
+    # Read a single JSON array from stdin
+    raw = sys.stdin.read().strip()
+    if not raw:
+        print("{}")
+        return
```

Comment:
should we say that we didn't get valid info here instead of just {}?

Reply:
Probably. I'll change that